Hadrons/DilutedNoise.hpp

```diff
@@ -332,6 +355,138 @@ bool DistillationNoise<FImpl>::mapEmpty(void) const
+    std::vector<unsigned char> hcomb = GridChecksum::sha256( scomb.c_str() , sizeof(char)*scomb.size() ); 
+    return GridChecksum::sha256_string(hcomb);
+}
+
+template <typename FImpl>
+void DistillationNoise<FImpl>::save(const std::string filename, const std::string distilname, const unsigned int traj)
+{
+#ifdef HAVE_HDF5
+    if(grid_->IsBoss())
+    {
+        std::string filestem = filename + "." + std::to_string(traj) + ".h5";
+        Hdf5Writer writer(filestem);
+        push(writer, distilname);
+        auto &group = writer.getGroup();
+
+        //metadata write
+        //hash
+        hsize_t         hash_dim = 1;
+        std::string shash = generateHash();
+        H5NS::DataSpace hash_dataspace(1, &hash_dim);
+        H5NS::DataType  hash_datatype(H5T_STRING,shash.size());
+        H5NS::Attribute attr_hash = group.createAttribute("NoiseHash", hash_datatype, hash_dataspace);
+        attr_hash.write(hash_datatype, shash.c_str() );
+        //dilution sizes t,l,s
+        std::vector<int> dil_sizes = {dilutionSize(Index::t),dilutionSize(Index::l),dilutionSize(Index::s)};
+        hsize_t dilsize_dim = dil_sizes.size();
+        H5NS::DataSpace dil_dataspace(1, &dilsize_dim);
+        H5NS::Attribute attr_dil = group.createAttribute("DilutionSizes",  Hdf5Type<int>::type(), dil_dataspace);
+        attr_dil.write(Hdf5Type<int>::type(), dil_sizes.data());
+        //nnoise
+        unsigned int nnoise = noise_.size();
+        hsize_t nnoise_dim = 1;
+        H5NS::DataSpace nnoise_dataspace(1, &nnoise_dim);
+        H5NS::Attribute attr_nnoise = group.createAttribute("Nnoise",  Hdf5Type<unsigned int>::type(), nnoise_dataspace);
+        attr_nnoise.write(Hdf5Type<unsigned int>::type(), &nnoise);
+        
+        //noise write
+        std::vector<hsize_t>    datadim = {static_cast<hsize_t>(noiseSize_)};
+        H5NS::DataSpace         dataspace(datadim.size(), datadim.data());
+        for(unsigned int i=0 ; i<noise_.size() ; i++)
+        {
+            H5NS::DataSet           dataset = group.createDataSet("Noise_"+std::to_string(i), Hdf5Type<Type>::type(), dataspace);
+            dataset.write(&noise_[i].front() , Hdf5Type<Type>::type(), dataspace, dataspace);
+        }
+    }
+#else
+    HADRONS_ERROR(Implementation, "all-to-all matrix I/O needs HDF5 library");
+#endif
+}
+
+template <typename FImpl>
+void DistillationNoise<FImpl>::loadNoise(const std::string filestem, const std::string distilname, const int nnoise)
+{
+#ifdef HAVE_HDF5
+    Hdf5Reader reader(filestem);
+    push(reader, distilname);
+    auto &group = reader.getGroup();
+
+    //noise
+    resize(nnoise);
+    for(unsigned int i=0 ; i<noise_.size() ; i++)
+    {
+        H5NS::DataSet dataset =  group.openDataSet("Noise_"+std::to_string(i));
+        dataset.read(&noise_[i].front(), Hdf5Type<Type>::type());
+    }
+
+#else
+    HADRONS_ERROR(Implementation, "all-to-all matrix I/O needs HDF5 library");
+#endif
+}
+
+template <typename FImpl>
```

Comment:
does it make sense to have a different bespoke read routine for each attribute? why not read the whole file at once, including attributes?

Reply:
It's because of the way Antonin wrote DistillationNoise. The data per see is loaded on the parent class DistillationNoise, but interpreting (buildMap_) the dilution sizes is done at the child class (e.g. dilution sizes goes to `interlacing_` in InterlacedDistillationNoise), so that I have a `load()` in the child class that calls the parent ones, and this `load()` is called by the corresponding load module. When saving this is not a problem because I can access the dilution sizes through dilutionSize(Index) which is virtual, so I can have just one function that does everything at the parent class.

Reply:
ok that makes sense. looks a bit cumbersome to me but seems fine.

Reply:
What I think I'm gonna do now that you mentioned is to merge loadNoise() and loadDilSizes in a single parent load() and the much simpler child load() will call it

Reply:
Just did